src/KeyExchange/DiffieHellmanGroup.php

```diff
@@ -0,0 +1,219 @@
+<?php
+
+namespace Amp\Ssh\KeyExchange;
+
+
+use function Amp\call;
+use Amp\Promise;
+use Amp\Ssh\Message\KeyExchangeCurveInit;
+use Amp\Ssh\Message\KeyExchangeCurveReply;
+use Amp\Ssh\Message\Message;
+use Amp\Ssh\Transport\BinaryPacketHandler;
+
+/**
+ * @internal
+ */
+class DiffieHellmanGroup implements KeyExchange
+{
+    const GROUP14 = 14;
+
+    const GROUP16 = 16;
+
+    const GROUP18 = 18;
+
+    /**
+     * @var int
+     */
+    private $group;
+    /**
+     * @var string
+     */
+    private $hash;
+
+    public function __construct(int $group, string $hash)
+    {
+        $this->group = $group;
+        $this->hash = $hash;
+    }
+
+    public function getName(): string
+    {
+        return 'diffie-hellman-group' . $this->group . '-' . $this->hash;
+    }
+
+    public function exchange(BinaryPacketHandler $handler): Promise
+    {
+        return call(function () use ($handler) {
+
+            $prime = str_replace(' ', '', $this->getPrime());
```

Comment:
can we avoid doing this ? Certainly small and certainly pointless optim but i would like to avoid doing stuff that can be precalculated here.

Reply:
yes